spring-cloud-gcp-starter-trace 4.1.1 WARN logs after fix #1568

After fix #1568 for issue #1554, in the spring-cloud-gcp-trace-sample (version 4.1.1) app logs, after making a first GET request to http://localhost:8080, I see the following WARN logs:
```
WARN  z.r.AsyncReporter$BoundedAsyncReporter   : Spans were dropped due to exceptions. All subsequent errors will be logged at FINE level.
WARN  z.r.AsyncReporter$BoundedAsyncReporter   : Dropped 8 spans due to AssertionError(com.google.protobuf.InvalidProtocolBufferException: Protocol message end-group tag did not match expected tag.)

java.lang.AssertionError: com.google.protobuf.InvalidProtocolBufferException: Protocol message end-group tag did not match expected tag.
	at zipkin2.reporter.stackdriver.StackdriverSender.parseTraceIdPrefixedSpan(StackdriverSender.java:220)
	at zipkin2.reporter.stackdriver.StackdriverSender.sendSpans(StackdriverSender.java:168)
	at zipkin2.reporter.AsyncReporter$BoundedAsyncReporter.flush(AsyncReporter.java:299)
	at zipkin2.reporter.AsyncReporter$Flusher.run(AsyncReporter.java:378)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.google.protobuf.InvalidProtocolBufferException: Protocol message end-group tag did not match expected tag.
	at com.google.protobuf.InvalidProtocolBufferException.invalidEndTag(InvalidProtocolBufferException.java:137)
	at com.google.protobuf.CodedInputStream$ArrayDecoder.checkLastTagWas(CodedInputStream.java:641)
	at com.google.protobuf.CodedInputStream$ArrayDecoder.readMessage(CodedInputStream.java:873)
	at com.google.devtools.cloudtrace.v2.Span$Builder.mergeFrom(Span.java:10357)
	at com.google.devtools.cloudtrace.v2.Span$Builder.mergeFrom(Span.java:10043)
	at com.google.protobuf.AbstractMessage$Builder.mergeFrom(AbstractMessage.java:418)
	at com.google.protobuf.AbstractMessage$Builder.mergeFrom(AbstractMessage.java:317)
	at com.google.protobuf.AbstractMessageLite$Builder.mergeFrom(AbstractMessageLite.java:222)
	at com.google.protobuf.AbstractMessage$Builder.mergeFrom(AbstractMessage.java:535)
	at zipkin2.reporter.stackdriver.StackdriverSender.parseTraceIdPrefixedSpan(StackdriverSender.java:218)
	... 4 common frames omitted
```

If I add back the exclusion of ZipkinAutoConfiguration,
```
/** Sample spring boot application. */
@AutoConfiguration
@SpringBootApplication(exclude = ZipkinAutoConfiguration.class)
public class Application implements WebMvcConfigurer {
```
I don't see the WARN logs anymore.

Expected would be to not see the warnings when the exclude is missing.

--- a/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilter.java
+++ b/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilter.java
@@ -23,6 +23,7 @@ import org.springframework.boot.autoconfigure.AutoConfigurationMetadata;
 /**
  * Exclude Spring Boot AutoConfiguration classes as they provide incompatible beans
  * when using Cloud Trace.
+ *
  * @since 4.1.2
  */
 public class TraceAutoConfigurationFilter implements AutoConfigurationImportFilter {

--- a/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilter.java
+++ b/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilter.java
@@ -20,6 +20,11 @@ import java.util.Set;
 import org.springframework.boot.autoconfigure.AutoConfigurationImportFilter;
 import org.springframework.boot.autoconfigure.AutoConfigurationMetadata;
 
+/**
+ * Exclude Spring Boot AutoConfiguration classes as they provide incompatible beans
+ * when using Cloud Trace.
+ * @since 4.1.2
+ */
 public class TraceAutoConfigurationFilter implements AutoConfigurationImportFilter {
 
   private static final Set<String> SHOULD_SKIP = Set.of(

--- a/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilter.java
+++ b/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilter.java
@@ -1,0 +1,39 @@
+/*
+ * Copyright 2017-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.autoconfigure.trace;
+
+import java.util.Set;
+import org.springframework.boot.autoconfigure.AutoConfigurationImportFilter;
+import org.springframework.boot.autoconfigure.AutoConfigurationMetadata;
+
+public class TraceAutoConfigurationFilter implements AutoConfigurationImportFilter {
+
+  private static final Set<String> SHOULD_SKIP = Set.of(
+      "org.springframework.boot.actuate.autoconfigure.tracing.zipkin.ZipkinAutoConfiguration");
+
+  @Override
+  public boolean[] match(String[] autoConfigurationClasses,
+      AutoConfigurationMetadata autoConfigurationMetadata) {
+    boolean[] matches = new boolean[autoConfigurationClasses.length];
+
+    for (int i = 0; i < autoConfigurationClasses.length; i++) {
+      matches[i] = autoConfigurationClasses[i] == null
+          || !SHOULD_SKIP.contains(autoConfigurationClasses[i]);
+    }
+    return matches;
+  }
+}

--- a/spring-cloud-gcp-autoconfigure/src/main/resources/META-INF/spring.factories
+++ b/spring-cloud-gcp-autoconfigure/src/main/resources/META-INF/spring.factories
@@ -10,3 +10,6 @@ com.google.cloud.spring.autoconfigure.secretmanager.SecretManagerConfigDataLocat
 # ConfigData Loaders
 org.springframework.boot.context.config.ConfigDataLoader=\
 com.google.cloud.spring.autoconfigure.secretmanager.SecretManagerConfigDataLoader
+
+org.springframework.boot.autoconfigure.AutoConfigurationImportFilter=\
+com.google.cloud.spring.autoconfigure.trace.TraceAutoConfigurationFilter

--- a/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/trace/StackdriverTraceAutoConfigurationTests.java
+++ b/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/trace/StackdriverTraceAutoConfigurationTests.java
@@ -17,6 +17,7 @@
 package com.google.cloud.spring.autoconfigure.trace;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.awaitility.Awaitility.await;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
@@ -51,7 +52,9 @@ import java.util.concurrent.TimeUnit;
 import org.apache.commons.lang3.reflect.FieldUtils;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.NoSuchBeanDefinitionException;
 import org.springframework.boot.actuate.autoconfigure.tracing.BraveAutoConfiguration;
+import org.springframework.boot.actuate.autoconfigure.tracing.zipkin.ZipkinAutoConfiguration;
 import org.springframework.boot.autoconfigure.AutoConfigurations;
 import org.springframework.boot.test.context.runner.ApplicationContextRunner;
 import org.springframework.cloud.autoconfigure.RefreshAutoConfiguration;
@@ -80,7 +83,8 @@ class StackdriverTraceAutoConfigurationTests {
                 StackdriverTraceAutoConfiguration.class,
                 GcpContextAutoConfiguration.class,
                 BraveAutoConfiguration.class,
-                RefreshAutoConfiguration.class))
+                RefreshAutoConfiguration.class,
+                TraceAutoConfigurationFilter.class))
         .withUserConfiguration(MockConfiguration.class)
         .withPropertyValues(
             "spring.cloud.gcp.project-id=proj");
@@ -101,6 +105,16 @@ class StackdriverTraceAutoConfigurationTests {
                   .isNotNull();
               assertThat(context.getBean(ManagedChannel.class)).isNotNull();
             });
+  }
+
+  @Test
+  void testDisableZipkinAutoConfiguration() {
+    this.contextRunner
+        .run(
+            context ->
+                assertThatThrownBy(() -> context.getBean(ZipkinAutoConfiguration.class))
+                    .isExactlyInstanceOf(NoSuchBeanDefinitionException.class)
+                    .hasMessageContaining("ZipkinAutoConfiguration"));
   }
 
   @Test

--- a/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilterTests.java
+++ b/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilterTests.java
@@ -1,33 +1,34 @@
 package com.google.cloud.spring.autoconfigure.trace;
 
-import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
 
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
-import org.springframework.beans.factory.NoSuchBeanDefinitionException;
-import org.springframework.boot.actuate.autoconfigure.tracing.zipkin.ZipkinAutoConfiguration;
-import org.springframework.boot.autoconfigure.AutoConfigurations;
-import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+import org.springframework.boot.autoconfigure.AutoConfigurationMetadata;
 
 class TraceAutoConfigurationFilterTests {
 
-  private ApplicationContextRunner contextRunner;
+  private TraceAutoConfigurationFilter filter;
 
   @BeforeEach
   void init() {
-    contextRunner = new ApplicationContextRunner()
-        .withConfiguration(
-            AutoConfigurations.of(
-                TraceAutoConfigurationFilter.class));
+    filter = new TraceAutoConfigurationFilter();
   }
 
   @Test
-  void test() {
-    this.contextRunner
-        .run(
-            context ->
-                assertThatThrownBy(() -> context.getBean(ZipkinAutoConfiguration.class))
-                    .isExactlyInstanceOf(NoSuchBeanDefinitionException.class)
-                    .hasMessageContaining("ZipkinAutoConfiguration"));
+  void testNullString() {
+    assertThat(filter.match(new String[1], mock(AutoConfigurationMetadata.class)))
+        .hasSize(1)
+        .contains(true);
+  }
+
+  @Test
+  void testZipkinAutoConfigurationShouldNotMatch() {
+    assertThat(filter.match(new String[]{
+        "org.springframework.boot.actuate.autoconfigure.tracing.zipkin.ZipkinAutoConfiguration"},
+        mock(AutoConfigurationMetadata.class)))
+        .hasSize(1)
+        .contains(false);
   }
 }

--- a/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilterTests.java
+++ b/spring-cloud-gcp-autoconfigure/src/test/java/com/google/cloud/spring/autoconfigure/trace/TraceAutoConfigurationFilterTests.java
@@ -1,0 +1,33 @@
+package com.google.cloud.spring.autoconfigure.trace;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.NoSuchBeanDefinitionException;
+import org.springframework.boot.actuate.autoconfigure.tracing.zipkin.ZipkinAutoConfiguration;
+import org.springframework.boot.autoconfigure.AutoConfigurations;
+import org.springframework.boot.test.context.runner.ApplicationContextRunner;
+
+class TraceAutoConfigurationFilterTests {
+
+  private ApplicationContextRunner contextRunner;
+
+  @BeforeEach
+  void init() {
+    contextRunner = new ApplicationContextRunner()
+        .withConfiguration(
+            AutoConfigurations.of(
+                TraceAutoConfigurationFilter.class));
+  }
+
+  @Test
+  void test() {
+    this.contextRunner
+        .run(
+            context ->
+                assertThatThrownBy(() -> context.getBean(ZipkinAutoConfiguration.class))
+                    .isExactlyInstanceOf(NoSuchBeanDefinitionException.class)
+                    .hasMessageContaining("ZipkinAutoConfiguration"));
+  }
+}

